### PR TITLE
RTMDet-tiny enablement for detection task

### PR DESCRIPTION
**Pull request by sungchul1**
### Summary

ticket : [CVS-142365](https://jira.devtools.intel.com/browse/CVS-142365)

This PR includes:
- [x] RTMDet-tiny enablement on detection task
- [x] Minor refactoring

Benchmark results on develop:

all | large | medium | small
-- | -- | -- | --
test/f1-score | test/f1-score | test/f1-score | test/f1-score
0.671 | 0.759 | 0.746 | 0.508

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have ran e2e tests and there is no issues.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/algo/detection/heads/delta_xywh_bbox_coder.py [hunk 1]
```diff
@@ -8,6 +8,8 @@
 import torch
 from torch import Tensor
 
+from otx.algo.detection.utils.utils import clip_bboxes_export
```
Comment (wonjuleee) [inline src/otx/algo/detection/heads/delta_xywh_bbox_coder.py:11]: Is there any special reason to rename this?

File: src/otx/algo/detection/utils/utils.py [hunk 2]
```diff
@@ -424,6 +425,71 @@ def distance2bbox(
     return bboxes
 
 
+def distance2bbox_export(points: Tensor, distance: Tensor, max_shape: Tensor | None = None) -> Tensor:
```
Comment (wonjuleee) [inline src/otx/algo/detection/utils/utils.py:428]: Can we add the argument `export=False` to `distance2bbox` and `clip_bboxes` for reusing the code blocks?

File: src/otx/recipe/detection/rtmdet_tiny.yaml [hunk 3]
```diff
@@ -0,0 +1,151 @@
+model:
+  class_path: otx.algo.detection.rtmdet.RTMDetTiny
+  init_args:
+    label_info: 80
+
+    optimizer:
+      class_path: torch.optim.AdamW
+      init_args:
+        lr: 0.0007
```
Comment (wonjuleee) [inline src/otx/recipe/detection/rtmdet_tiny.yaml:9]: how this number was chosen?

File: src/otx/algo/detection/heads/delta_xywh_bbox_coder.py (hunk 1 above)
Comment (sungchul1) [reply src/otx/algo/detection/heads/delta_xywh_bbox_coder.py:11]: In the data pipeline, there is `clip_bboxes` as well.
I renamed this function to distinguish functions for the next step, refactoring duplicated modules.

File: src/otx/algo/detection/utils/utils.py (hunk 2 above)
Comment (sungchul1) [reply src/otx/algo/detection/utils/utils.py:428]: Thanks for the recommendation.
I'll do it when refactoring duplicated functions.

File: src/otx/recipe/detection/rtmdet_tiny.yaml [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+              random_pop: false
+          - class_path: otx.core.data.transform_libs.torchvision.RandomResize
+            init_args:
+              scale:
+                - 1280
+                - 1280
+              ratio_range:
+                - 0.5
+                - 2.0
+              keep_ratio: true
```
Comment (wonjuleee) [inline src/otx/recipe/detection/rtmdet_tiny.yaml:59]: could you clarify the relationship between ratio_range and keep_ratio?

File: src/otx/recipe/detection/rtmdet_tiny.yaml (hunk 3 above)
Comment (sungchul1) [reply src/otx/recipe/detection/rtmdet_tiny.yaml:9]: I used HPO with optimizer (SGD, AdamW) and learning rate on 4 benchmark datasets (bdd medium/large, pothole medium, vitens large).
I got 8 hyper parameter candidates and 1 official hyper parameter, I did additional benchmark experiments to finalize the hyper parameter.
This number showed the best performance with shorter epochs.

File: src/otx/recipe/detection/rtmdet_tiny.yaml (hunk 4 above)
Comment (sungchul1) [reply src/otx/recipe/detection/rtmdet_tiny.yaml:59]: `ratio_range` is to decide what `scale` will be used, and `keep_ratio` is to keep the aspect ratio when resizing the image.
For example, if `ratio_range` is 0.5, `scale` will be 640 (=1280 * 0.5). If a given image is (160, 320), the image will become (320, 640).

File: src/otx/algo/detection/heads/distance_point_bbox_coder.py [hunk 5]
```diff
@@ -84,3 +84,18 @@ def decode(
...
         return distance2bbox(points, pred_bboxes, max_shape)
+
+    def decode_export(
+        self,
+        points: Tensor,
+        pred_bboxes: Tensor,
+        max_shape: tuple[int, ...] | Tensor | tuple[tuple[int, ...], ...] | None = None,
+    ) -> Tensor:
+        """Decode distance prediction to bounding box for export."""
+        assert points.size(0) == pred_bboxes.size(0)  # noqa: S101
```
Comment (eunwoosh) [inline src/otx/algo/detection/heads/distance_point_bbox_coder.py]: Is there any reason to use assert instead of raise Error?

File: src/otx/algo/detection/heads/rtmdet_head.py [hunk 6]
```diff
@@ -152,11 +166,28 @@ def loss_by_feat_single(  # type: ignore[override]
...
+            bbox_targets (Tensor): BBox regression targets of each anchor with
+                shape (N, num_total_anchors, 4).
+            assign_metrics (Tensor): Assign metrics with shape
+                (N, num_total_anchors).
+            stride (list[int]): Downsample stride of the feature map.
+
+        Returns:
+            dict[str, Tensor]: A dictionary of loss components.
+        """
+        assert stride[0] == stride[1], "h stride is not equal to w stride!"  # noqa: S101
```
Comment (eunwoosh) [inline src/otx/algo/detection/heads/rtmdet_head.py]: Same as above

File: src/otx/algo/detection/heads/rtmdet_head.py [hunk 7]
```diff
@@ -452,3 +548,175 @@ def get_anchors(
...
+    def __init__(
+        self,
+        num_classes: int,
+        in_channels: int,
+        share_conv: bool = True,
+        use_depthwise: bool = False,
+        norm_cfg: dict | None = None,
+        act_cfg: dict | None = None,
+        pred_kernel_size: int = 1,
+        exp_on_reg: bool = False,
```
Comment (eunwoosh) [inline src/otx/algo/detection/heads/rtmdet_head.py:598]: Could you add docstring for this argument?

File: src/otx/algo/detection/rtmdet.py [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+            bboxes=bboxes,
+            labels=labels,
+        )
+
+    def get_classification_layers(self, prefix: str = "") -> dict[str, dict[str, int]]:
+        """Return classification layer names by comparing two different number of classes models.
+
+        Args:
+            config (DictConfig): Config for building model.
+            model_registry (Registry): Registry for building model.
```
Comment (eunwoosh) [inline src/otx/algo/detection/rtmdet.py]:
```suggestion
```
It seems unnecessary.

**Review by eunwoosh — COMMENTED** [review]
Thanks for your work :) I left some comments. Please take a look.

File: src/otx/algo/detection/heads/distance_point_bbox_coder.py (hunk 5 above)
Comment (sungchul1) [reply src/otx/algo/detection/heads/distance_point_bbox_coder.py]: No reason. I got mmdeploy source as it is and ignored ruff error.

File: src/otx/algo/detection/rtmdet.py (hunk 8 above)
Comment (sungchul1) [reply src/otx/algo/detection/rtmdet.py]: Done at https://github.com/openvinotoolkit/training_extensions/pull/3542/commits/61589d90eebb3a3d82c0efa4ac37d7772b4b00f2.
Thanks for finding! I modified instance segmentation parts as well.

File: src/otx/algo/detection/heads/rtmdet_head.py (hunk 7 above)
Comment (sungchul1) [reply src/otx/algo/detection/heads/rtmdet_head.py:598]: Done at https://github.com/openvinotoolkit/training_extensions/pull/3542/commits/589b361727fb74be22273eb9aaa6dab48133760b.

File: src/otx/algo/detection/heads/rtmdet_head.py (hunk 6 above)
Comment (sungchul1) [reply src/otx/algo/detection/heads/rtmdet_head.py]: No reason. I got mmdeploy source as it is and ignored ruff error.

File: src/otx/algo/detection/heads/distance_point_bbox_coder.py (hunk 5 above)
Comment (eunwoosh) [reply src/otx/algo/detection/heads/distance_point_bbox_coder.py]: Could you change all of assert to if... raise? As you know, `assert` code is removed if python is requested to optimize. I think using if... raise is better than assert.

Comment (sungchul1) [reply src/otx/algo/detection/heads/distance_point_bbox_coder.py]: Done at https://github.com/openvinotoolkit/training_extensions/pull/3542/commits/d7a4e92ff4ddef6b226fa00543fa24fbc8233cfd.

File: src/otx/algo/detection/heads/rtmdet_head.py (hunk 6 above)
Comment (sungchul1) [reply src/otx/algo/detection/heads/rtmdet_head.py]: Done at https://github.com/openvinotoolkit/training_extensions/pull/3542/commits/d7a4e92ff4ddef6b226fa00543fa24fbc8233cfd.

**Review by eunwoosh — APPROVED** [review]

**Review by wonjuleee — APPROVED** [review]
